src/asset_manager/asset.rs

```diff
@@ -0,0 +1,200 @@
+///         let mut num_columns = 0;
+///         let mut old_columns = 0;
+///         let mut values = Vec::new();
+///         let mut reader = bytes.as_ref();
+///
+///         for line in reader.lines() {
+///             let line = line?;
+///
+///             for word in line.split_whitespace() {
+///                 num_columns += 1;
+///                 values.push(word.parse().map_err(|x| ImportError::FormatError("Invalid integer".to_string()))?)
+///             }
+///             if old_columns != 0 && num_columns != old_columns {
+///                 return Err(ImportError::FormatError(format!("Number of columns not matching (previous line: {}, this line: {})",
+///                                                             old_columns, num_columns)));
+///             }
+///             old_columns = num_columns;
+///             num_columns = 0;
+///         }
+///
+///         if values.len() == 0 {
+///             return Err(ImportError::FormatError("Empty file".to_string()));
+///         }
+///
+///         Ok(Table {
+///             num_columns: old_columns,
+///             num_rows: values.len() as u32 / old_columns,
+///             data: values.into_boxed_slice(),
+///         })
+///     }
+/// }
+/// ```
+pub trait Asset: Sized {
+    /// The data type, an intermediate format.
+    /// This may also be `Self` if this asset does
+    /// not depend on `Context` (if it is just
+    /// plain data).
+    type Data;
+    /// The error type that may be returned if
+    /// `from_data` fails.
+    type Error: Debug;
+
+    /// Returns the category, which is
+    /// used for subfolders.
+    ///
+    /// Examples are `"meshes"` and `"textures"`.
+    fn category() -> &'static str;
+
+    /// Create the asset from the data and the context (used to create buffers for the gpu).
+    fn from_data(data: Self::Data, context: &mut Context) -> Result<Self, Self::Error>;
+}
+
+/// Specifies an asset format. Note that
+/// the asset format is not the same as the type.
+/// There may be multiple formats for the same
+/// asset. A format should also implement the
+/// `Import` trait for the target `Asset`s data
+/// type.
+pub trait AssetFormat {
+    /// Return the typical file extensions a file
+    /// with this format has, without the preceding `"."`.
+    /// If there is no asset with the first extension, the
+    /// next ones will be tested for existence.
+    ///
+    /// # Panics
+    ///
+    /// May cause a panic if the length of
+    /// file extensions is 0! Do not do this.
+    fn file_extensions() -> &'static [&'static str];
+}
+
+/// An asset store may be a ".zip" file, a server,
+/// a custom binary format conataining levels, or just
+/// a directory.
+pub trait AssetStore {
+    /// Read an asset from a given name and format and
+    /// return the bytes.
+    fn read_asset<T, F>(&self, name: &str, format: &F) -> Result<Box<[u8]>, AssetStoreError>
```

Comment:
Minor: Can remove `format: &F` here imo, because `AssetFormat` only has `file_extensions()`, which is static.